python3/libexec/usb_scan.py

```diff
@@ -374,7 +374,7 @@ def __init__(self):
         """
         self.rule_list = []
         try:
-            with open(self._PATH) as f:
+            with open(self._PATH, encoding="utf-8", errors="ignore") as f:
```

Comment:
With `errors="ignore"`, will the encoding related errors be hidden?
I see the [description](https://docs.python.org/3/library/functions.html#open) for this parameter:
`'ignore' ignores errors. Note that ignoring encoding errors can lead to data loss.
`

Reply:
there is a comment about this change. https://github.com/xapi-project/xen-api/pull/5424#pullrequestreview-1931214537

Reply:
Thanks for the link to the previous discussion~
Then in case there are invalid utf-8 encoding sequences, those sequences will just be ignored so this piece of code will still run without any error, is my understanding right? Thank you!

Reply:
I'm waiting response before approving

Reply:
> I'm waiting response before approving
Yes, this is what I referred to in https://github.com/xapi-project/xen-api/pull/5424#pullrequestreview-1931214537, here is an example of it working:
```py
echo -e '<before>\0345<after>' >datafile.txt
python3.6 -c 'print(open("datafile.txt",encoding="utf-8",errors="ignore" ).read())'
<before><after>
python3.6 -c 'print(open("datafile.txt",encoding="utf-8",errors="replace").read())'
<before>�<after>
# vs
python3.6 -c 'print(open("datafile.txt",encoding="utf-8"                 ).read())'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/usr/lib64/python3.6/codecs.py", line 321, in decode
    (result, consumed) = self._buffer_decode(data, self.errors, final)
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xe5 in position 8: invalid continuation byte
```
BTW, The same applies to `b"bytes-with-unclear-content".decode()` (and also to `Popen`):
```py
python3.6 -c 'print(b"<before>\345<after>".decode("utf-8", "ignore"           ))'
<before><after>
python3.6 -c 'print(b"<before>\345<after>".decode("utf-8", "backslashreplace" ))'
<before>\xe5<after>
python3.6 -c 'print(b"<before>\345<after>".decode("utf-8"                     ))'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xe5 in position 8: invalid continuation byte


```
There are various other options when a program opens a file in text (not binary) mode, where an automatic decode() has to happen as part of `read()`: https://docs.python.org/3.6/library/functions.html#open

> errors is an optional string that specifies how encoding and decoding errors are to be handled—this cannot be used in binary mode. A variety of standard error handlers are available (listed under [Error Handlers](https://docs.python.org/3.6/library/codecs.html#error-handlers)), though any error handling name that has been registered with [codecs.register_error()](https://docs.python.org/3.6/library/codecs.html#codecs.register_error) is also valid. The standard names include:
>    - 'strict' to raise a [ValueError](https://docs.python.org/3.6/library/exceptions.html#ValueError) exception if there is an encoding error. The default value of None has the same effect.
>    - 'ignore' ignores errors. Note that ignoring encoding errors can lead to data loss.
>    - 'replace' causes a replacement marker (such as '?') to be inserted where there is malformed data.
>    - 'surrogateescape' will represent any incorrect bytes as code points in the Unicode Private Use Area ranging from U+DC80 to U+DCFF. These private code points will then be turned back into the same bytes when the surrogateescape error handler is used when writing data. This is useful for processing files in an unknown encoding.
>    - 'xmlcharrefreplace' is only supported when writing to a file. Characters not supported by the encoding are replaced with the appropriate XML character reference &#nnn;.
>    - 'backslashreplace' replaces malformed data by Python’s backslashed escape sequences.
>    - 'namereplace' (also only supported when writing) replaces unsupported characters with \N{...} escape sequences.

Tthe other options besides `errors="ignore"` deal with various ways to preserve the bytes that would otherwise trigger `UnicodeDecodeError` as shown above.

If you expect such non-UTF8, non-ASCCI bytes in a file and want to handle them, then you can choose among these other options. But if you don't expect them, using `errors="ignore"` is enough to at least not raise an `UnicodeDecodeError` when reading the file.

Of course, if a file is a well-known plain ASCII file that can never have any non-ASCII characters (like we'd expect from many `sysfs` attributes in `/sys` and many files in `/proc`). In these cases, it is of course safe to ignore the warning for that specific `open()`. Or if there is more than one such file used in a Python module the entire file (unless other `open`'s are added to the file).

In the specific comment in https://github.com/xapi-project/xen-api/pull/5424#pullrequestreview-1931214537, I was just asked about the generic safety of not specifying and encoding parameter open opening files in test mode, and there, I had to answer that it's in generically safe to always not specify and encoding, and to be safe against random non-ASCII data, an `errors=` that is set to something else than the default of `errors="strict"` that leads to an `UnicodeDecodeError` in such rare, but unfortunate cases.

Summary:

- Yes, this indeed protects against unexpected decoding errors.
- Yes, in the case of `errors="ignore"` such non-UTF8 bytes are discarded by `read()` in this case.
- So if you want to do something else, like copying the data to another output, or logging its contents in full, with unknown bytes shown in some transformed way, then you might want choose another `errors=` value.
  - If doing that, and you want to preserve the data fully, you've two options:
    - Convert all uses of the variable to work with bytes and open the file in binary mode
    - Use `errors="surrogateescape"`, which allows to encode() the bytes back to their original values when encoding for writing. I haven't tested this, but the Python manual on `open()` liked above says that it would be possible. 
  - However, unless you are doing one of those, `encoding="ignore"` is enough.
  - If the non-UTF8 bytes shall be shown in some format, e.g. in a log file, then one of the other `error=` that end in `replace` might be best, but not all of them work correctly with `read()`.
  - I only encountered a program that sent a "Latin" character in `latin1`/`iso-8859-1` character set encoding in its output that then failed to be read by `Popen()` when using text mode. In my case, I had to switch to processing the data as bytes because other libraries like (`tarfile` and `zipfile`) also work with bytes, so in this case the needed fix was to switch all code to use `bytes` instead, and using binary mode side-steps the entire encoding/decoding step completely by not converting to and from Unicode `str`.

To log such non-UTF-8 bytes, `errors="backslashreplace"` can be used:
```py
python3.6 -c 'print(open("datafile.txt",encoding="utf-8",errors="backslashreplace").read())'
<before>\xe5<after>
```
Of course, this turns each such non-UTF-8 byte into 4 bytes on output, so YMMV depending on how much data is being processed.
  
`tl;dr`:

As far as my thoughts went so far:
- If you don't want to log (or show) non-UTF8 bytes, but intead to make the code safe against `UnicodeDecodeError` exceptions being raised, what I see above like this here would be fine:
```py
 with open(self._PATH, encoding="utf-8", errors="ignore") as f:
```

Not knowing the context which file is read and what its content may or may not be, I would assume in most uses, I think this is enough to be safe against such Exceptions raised by the decoder.

Reply:
@bernhardkaindl Thank you very much for your help!

Reply:
The file is indeed tried to be printed, this should not be using ignore :)

Reply:
OK, will remove ignore here.